website/blog/2022-08-17-managing-surrogate-keys-in-dbt.md

```diff
@@ -0,0 +1,228 @@
+    
+        create sequence if not exists {{ model.database }}.{{ model.schema }}.{{ model.name }}_seq;
+        
+        {% endif %}
+    
+    {% endfor %}
+  
+{% endmacro %}
+
+```
+
+You can see in the above macro that we’re baking a naming convention here — for any model, the name of the sequence will exist in the same database and schema and follow the naming convention `<model_name>_seq`. Adhering to this pattern allows us to also create an easy macro to increment the sequences in our model definitions without having to hard code the sequence name in every model that needs a surrogate key. 
+
+```yaml
+# in macros/increment_sequence.sql
+
+{%- macro increment_sequence() -%}
+  
+  {{ this.name }}_seq.nextval
+
+{%- endmacro -%}
+```
+
+So your model code looks like:
+
+```yaml
+# in dim_customers
+...
+
+with cte_name as (
+		...
+)
+...
+
+select
+
+	{{ increment_sequence() }} as customer_id, 
+	first_name, 
+	last_name
+
+from cte_name
+
+...
+```
+
+### Caveats
+
+Despite the relative simplicity of this strategy, there are a handful of drawbacks with regard to making sure these sequences work the way we want them to. 
+
+- **dbt Run errors -** If an incremental model that has surrogate keys maintained in this way *fails* due to some SQL error, we may end up with gaps in our surrogate key. When dbt goes to execute the model, the sequence is queried, and therefore incremented, but the model failure prevents changes to the target table model. That means the next time we run the model, the incremental model will start on the wrong value, and we may end up with a column that looks like this:
+    
+    | surrogate_key_id | 
+    | --- | 
+    | 1 | 
+    | 2 | 
+    | 3 | 
+    | 4 | 
+    | 5 | x
+    | 8 | 
+    
+    In fact, most cloud platforms [can’t guarantee](https://docs.snowflake.com/en/user-guide/querying-sequences.html#:~:text=Snowflake%20does%20not%20guarantee%20generating%20sequence%20numbers%20without%20gaps.%20The%20generated%20numbers%20consistently%20increase%20in%20value%20(or%20decrease%20in%20value%20if%20the%20step%20size%20is%20negative)%20but%20are%20not%20necessarily%20contiguous) that sequences will be generated without gaps because of their use of parallel processing, even if we *don’t* have a dbt run error — because queries will be spread across multiple compute clusters, each step might query the sequence at different times, which makes it possible to have an out of order sequence result. This is a major consideration in using sequences — if that’s a deal breaker, you may need additional SQL logic in our models (like a `row_number()` function) to guarantee your keys are monotonically increasing. 
+    
+- **Views -** Because sequences in Snowflake increment on every query, using them as the surrogate keys for views would mean every time the view is queried, the sequence would increment and therefore change. This strategy would only work for table or incremental models.
```

Comment:
```suggestion
- **<Term id="view">Views</Term> -** Because sequences in Snowflake increment on every query, using them as the surrogate keys for views would mean every time the view is queried, the sequence would increment and therefore change. This strategy would only work for table or incremental models.
```

I'm not sure if this will cause wonkiness with the bold formatting, so recommend playing around with this locally, but suggestion is to link to View glossary page!